lib/route_matcher.rb

```diff
@@ -82,4 +84,20 @@ def format_allowed?(request)
     request_format = request.formats&.first&.symbol
     formats.include?(request_format)
   end
+
+  def path_params_from_request(request)
+    if request.env[ActionDispatch::Http::Parameters::PARAMETERS_KEY].nil?
```

Comment:
Just curious, will it be possible to avoid referencing the constant here because I'm not sure if this constant is even considered public. I believe it isn't because the `request.path_parameters` method is the way to access the header.

Reply:
It depends. Using `request.env[ActionDispatch::Http::Parameters::PARAMETERS_KEY]` is an optimization, because it's `nil` when it hasn't been set yet. And that's the only reason why we should try to use `recognize_path`. `request.path_parameters` always returns a Hash. It's an empty an Hash if the parameters couldn't be recognized, but at that point there's no way to tell why the Hash is empty. It might be a theoretical problem, but I wanted to run `recognize_path` only when Rails didn't try to do it yet.
So, yes, we could stop using that constant, remove that micro-optimization and completely rely on `request.path_parameters`.